lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchEvaluatorShimManager.java

```diff
@@ -0,0 +1,339 @@
+            .setEvaluatorLaunchCommand(command)
+            .setEvaluatorConfigString(evaluatorConfigurationString)
+            .build());
+    this.updateRuntimeStatus();
+  }
+
+  public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
+
+    synchronized (this.activeResources) {
+      String remoteId = this.activeResources.get(resourceReleaseEvent.getIdentifier());
+
+      if (!this.activeResources.containsKey(resourceReleaseEvent.getIdentifier())) {
+        LOG.log(Level.WARNING, "Received a ResourceReleaseEvent for an unknown resource id = {0}.",
+            resourceReleaseEvent.getIdentifier());
+      } else {
+        EventHandler<EvaluatorShimProtocol.EvaluatorShimControlProto> handler =
+            AzureBatchEvaluatorShimManager.this.remoteManager.getHandler(remoteId,
+                EvaluatorShimProtocol.EvaluatorShimControlProto.class);
+
+        LOG.log(Level.INFO, "Sending a TERMINATE command to the evaluator shim with remoteId = {0}.", remoteId);
+        handler.onNext(
+            EvaluatorShimProtocol.EvaluatorShimControlProto
+                .newBuilder()
+                .setCommand(EvaluatorShimProtocol.EvaluatorShimCommand.TERMINATE)
+                .build());
+      }
+
+      this.activeResources.remove(resourceReleaseEvent.getIdentifier());
+    }
+
+    this.updateRuntimeStatus();
+  }
+
+  public void onClose() {
+    try {
+      evaluatorShimCommandChannel.close();
+    } catch (Exception e) {
+      LOG.log(Level.WARNING, "An unexpected exception while closing the Evaluator Shim Command channel: {0}", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  private void updateRuntimeStatus() {
+    this.reefEventHandlers.onRuntimeStatus(RuntimeStatusEventImpl.newBuilder()
+        .setName(RUNTIME_NAME)
+        .setState(State.RUNNING)
+        .setOutstandingContainerRequests(this.outstandingResourceRequests.size())
+        .build());
+  }
+
+  private String getEvaluatorShimLaunchCommand() {
+    return this.launchCommandBuilder.buildEvaluatorShimCommand(EVALUATOR_SHIM_MEMORY_MB,
+        this.reefFileNames.getLocalFolderPath() + "/" + EVALUATOR_SHIM_CONF_FILE);
+  }
+
+  private FileResource getFileResourceFromFile(final File configFile, final FileType type) {
+    return FileResourceImpl.newBuilder()
+        .setName(configFile.getName())
+        .setPath(configFile.getPath())
+        .setType(type).build();
+  }
+
+  private Configuration buildShimConfiguration(final String azureBatchTaskId) {
+    return EvaluatorShimConfiguration.CONF
+        .set(EvaluatorShimConfiguration.DRIVER_REMOTE_IDENTIFIER, this.remoteManager.getMyIdentifier())
+        .set(EvaluatorShimConfiguration.CONTAINER_IDENTIFIER, azureBatchTaskId)
+        .build();
+  }
+
+  private File writeShimConfigurationFile(final String azureBatchTaskId) {
+    final File shimConfigurationFile = new File(this.reefFileNames.getLocalFolderPath()
+        + "/" + EVALUATOR_SHIM_CONF_FILE);
```

Comment:
Paths.Get or File.separatorChar?

Reply:
Hm, they're using "/" all the time in REEFFileNames. But that, of course, doesn't necessarily mean that's correct. So I'll change it.